x-pack/plugins/security_solution/public/detection_engine/rule_creation_ui/components/data_view_selector/data_view_selector.tsx

```diff
@@ -6,85 +6,64 @@
-  kibanaDataViews: Record<string, DataViewListItem>;
   field: FieldHook<DefineStepRule['dataViewId']>;
 }
 
-export const DataViewSelector = ({ kibanaDataViews, field }: DataViewSelectorProps) => {
-  let isInvalid;
-  let errorMessage;
-  let dataViewId: string | null | undefined;
+export const DataViewSelector = ({ field }: DataViewSelectorProps) => {
+  const kibanaDataViews = useDataViews();
+  const dataViewId: DataViewId | null | undefined = field?.value;
+  const fieldAndError = field ? getFieldValidityAndErrorMessage(field) : undefined;
+  const isInvalid = fieldAndError?.isInvalid;
+  const errorMessage = fieldAndError?.errorMessage;
 
-  if (field != null) {
-    const fieldAndError = getFieldValidityAndErrorMessage(field);
-    isInvalid = fieldAndError.isInvalid;
-    errorMessage = fieldAndError.errorMessage;
-    dataViewId = field.value;
-  }
+  const selectedKibanaDataView = useMemo(() => {
+    if (!isDataViewIdValid(dataViewId)) {
+      return;
+    }
 
-  const kibanaDataViewsDefined = useMemo(
-    () => kibanaDataViews != null && Object.keys(kibanaDataViews).length > 0,
-    [kibanaDataViews]
-  );
+    return kibanaDataViews?.find((x) => x.id === dataViewId);
+  }, [dataViewId, kibanaDataViews]);
 
-  // Most likely case here is that a data view of an existing rule was deleted
-  // and can no longer be found
-  const selectedDataViewNotFound = useMemo(
-    () =>
-      dataViewId != null &&
-      dataViewId !== '' &&
-      kibanaDataViewsDefined &&
-      !Object.hasOwn(kibanaDataViews, dataViewId),
-    [kibanaDataViewsDefined, dataViewId, kibanaDataViews]
-  );
   const [selectedOption, setSelectedOption] = useState<Array<EuiComboBoxOptionOption<string>>>(
-    !selectedDataViewNotFound && dataViewId != null && dataViewId !== ''
-      ? [{ id: kibanaDataViews[dataViewId].id, label: kibanaDataViews[dataViewId].title }]
+    selectedKibanaDataView
+      ? [{ id: selectedKibanaDataView.id, label: selectedKibanaDataView.title }]
       : []
   );
-
   const [showDataViewAlertsOnAlertsWarning, setShowDataViewAlertsOnAlertsWarning] = useState(false);
 
   useEffect(() => {
-    if (!selectedDataViewNotFound && dataViewId) {
-      const dataViewsTitle = kibanaDataViews[dataViewId].title;
-      const dataViewsId = kibanaDataViews[dataViewId].id;
-
-      setShowDataViewAlertsOnAlertsWarning(dataViewsId === 'security-solution-default');
-
-      setSelectedOption([{ id: dataViewsId, label: dataViewsTitle }]);
-    } else {
+    if (!selectedKibanaDataView) {
       setSelectedOption([]);
+      return;
     }
-  }, [
-    dataViewId,
-    field,
-    kibanaDataViews,
-    selectedDataViewNotFound,
-    setShowDataViewAlertsOnAlertsWarning,
-  ]);
 
-  // TODO: optimize this, pass down array of data view ids
-  // at the same time we grab the data views in the top level form component
+    const dataViewsId = selectedKibanaDataView.id;
+    const dataViewsTitle = selectedKibanaDataView.title;
```

Comment:
IMO assigning these to variables doesn't make this code any clearer; `dataViewsId` doesn't have any inherent meaning.

I _think_ this could be changed to look like:

```suggestion
    const { id, value } = formDataView;
```

where `selectedKibanaDataView` has been renamed to `formDataView`, as it's the data view derived from `field.value` and not truly the "selected" data view from the UI' perspective. I think that would make this `useEffect` a little more clear, as well: its purpose is to reflect the form value in the UI, by setting the selected option to `formDataView`.

Reply:
I further simplified the component removed unnecessary code chunks. Hopefully it's much more readable now.